### PR TITLE
#497 multistep

### PR DESCRIPTION
**Pull request by even1024**
Implements multistep feature #497

### CODE REVIEW

File: core/indigo-core/common/math/algebra.h [hunk 1]
```diff
@@ -303,6 +303,54 @@ namespace indigo
             _rightTop = other._rightTop;
         }
 
+        inline bool pt_in_rect(const Vec2f& pt) const
```
Comment (SunFellow) [inline core/indigo-core/common/math/algebra.h:306]: Please avoid abbreviations in method names.
Looking at other names, here was used camelCase style, not Snake case. Let's follow the style of the header.

File: core/indigo-core/common/math/algebra.h [hunk 2]
```diff
@@ -303,6 +303,54 @@ namespace indigo
...
+        {
+            return _leftBottom.x < pt.x && _leftBottom.y < pt.y && _rightTop.x > pt.x && _rightTop.y > pt.y;
+        }
+
+        inline bool ray_intersects_rect(const Vec2f& begin, const Vec2f& end)
```
Comment (SunFellow) [inline core/indigo-core/common/math/algebra.h:311]: Looking at other names, here was used camelCase style, not Snake case. Let's follow the style of the header.

File: core/indigo-core/common/math/algebra.h [hunk 3]
```diff
@@ -303,6 +303,54 @@ namespace indigo
...
+            bool same_sign = std::signbit(Vec2f::dot(vr, lb)) == std::signbit(Vec2f::dot(vr, lt)) &&
+                             std::signbit(Vec2f::dot(vr, lt)) == std::signbit(Vec2f::dot(vr, rt)) &&
+                             std::signbit(Vec2f::dot(vr, rt)) == std::signbit(Vec2f::dot(vr, rb));
+
+            return !same_sign && Vec2f::cross(lb, vr) > 0 && Vec2f::cross(lt, vr) > 0 && Vec2f::cross(rt, vr) > 0 && Vec2f::cross(rb, vr) > 0;
+        }
+
+#define HYPOT(a, b) (sqrt((a) * (a) + (b) * (b)))
+
+        inline double pt_distance(const Vec2f& pt)
```
Comment (SunFellow) [inline core/indigo-core/common/math/algebra.h:329]: Looking at other names, here was used camelCase style, not Snake case. Let's follow the style of the header.

File: core/indigo-core/common/math/algebra.h (hunk 1 above)
Comment (SunFellow) [inline core/indigo-core/common/math/algebra.h:306]: Let's follow one naming style. The file uses a mix of camelCase and Snake case styles but majority of methods follow camelCase.

File: core/indigo-core/graph/graph.h [hunk 4]
```diff
@@ -125,18 +125,18 @@ namespace indigo
...
+    class MetaObject
     {
     public:
-        GraphMetaObject(int class_id) : _class_id(class_id)
+        MetaObject(int class_id) : _class_id(class_id)
         {
         }
-        int _class_id;
-        virtual GraphMetaObject* clone() const = 0;
+        uint32_t _class_id;
```
Comment (SunFellow) [inline core/indigo-core/graph/graph.h:136]: It makes sense to keep coordinated types of constructor parameter and data member.

File: core/indigo-core/molecule/base_molecule.h [hunk 5]
```diff
@@ -466,6 +484,7 @@ namespace indigo
         // When molecule gets edited then edit revision is increased.
         // If edit revision is the same then molecule wasn't edited
         int _edit_revision;
+        PtrArray<MetaObject> _meta_data;
```
Comment (SunFellow) [inline core/indigo-core/molecule/base_molecule.h:487]: We have a chance to get rid of hand-made PtrArray ;)

Comment (even1024) [reply core/indigo-core/molecule/base_molecule.h:487]: We can't do it in BaseMolecule class.

File: core/indigo-core/graph/graph.h [hunk 6]
```diff
@@ -125,18 +125,18 @@ namespace indigo
...
     };
 
-    class GraphMetaObject
+    class CycleBasis;
+
+    class MetaObject
     {
     public:
-        GraphMetaObject(int class_id) : _class_id(class_id)
+        MetaObject(int class_id) : _class_id(class_id)
```
[the remark this answers is not in the document]
Comment (even1024) [reply core/indigo-core/graph/graph.h]: fixed

File: core/indigo-core/graph/graph.h (hunk 4 above)
Comment (even1024) [reply core/indigo-core/graph/graph.h]: fixed

File: core/indigo-core/graph/graph.h [hunk 7]
```diff
@@ -125,18 +125,18 @@ namespace indigo
...
         {
         }
-        int _class_id;
-        virtual GraphMetaObject* clone() const = 0;
+        uint32_t _class_id;
+        virtual MetaObject* clone() const = 0;
     };
```
[the remark this answers is not in the document]
Comment (even1024) [reply core/indigo-core/graph/graph.h]: fixed

File: core/indigo-core/common/math/algebra.h (hunk 1 above)
Comment (even1024) [reply core/indigo-core/common/math/algebra.h]: fixed

File: core/indigo-core/common/math/algebra.h (hunk 3 above)
Comment (even1024) [reply core/indigo-core/common/math/algebra.h]: fixed

File: core/indigo-core/common/math/algebra.h (hunk 2 above)
Comment (even1024) [reply core/indigo-core/common/math/algebra.h]: fixed

File: core/indigo-core/common/math/algebra.h (hunk 1 above)
Comment (even1024) [reply core/indigo-core/common/math/algebra.h]: fixed

File: core/indigo-core/common/math/algebra.h [hunk 8]
```diff
@@ -303,6 +303,54 @@ namespace indigo
...
+            auto rb = rightBottom() - begin;
+            // same_sign means no intersection
+            bool same_sign = std::signbit(Vec2f::dot(vr, lb)) == std::signbit(Vec2f::dot(vr, lt)) &&
+                             std::signbit(Vec2f::dot(vr, lt)) == std::signbit(Vec2f::dot(vr, rt)) &&
+                             std::signbit(Vec2f::dot(vr, rt)) == std::signbit(Vec2f::dot(vr, rb));
+
+            return !same_sign && Vec2f::cross(lb, vr) > 0 && Vec2f::cross(lt, vr) > 0 && Vec2f::cross(rt, vr) > 0 && Vec2f::cross(rb, vr) > 0;
+        }
+
+#define HYPOT(a, b) (sqrt((a) * (a) + (b) * (b)))
```
[the remark this answers is not in the document]
Comment (even1024) [reply core/indigo-core/common/math/algebra.h]: fixed

File: core/indigo-core/molecule/ket_commons.h [hunk 9]
```diff
@@ -36,17 +37,17 @@ namespace indigo
         return string_hash(s, count);
     }
 
-    class KETSimpleObject : public GraphMetaObject
+    class KETSimpleObject : public MetaObject
     {
     public:
         static const std::uint32_t cid = "KET simple object"_hash;
```
Comment (SunFellow) [inline core/indigo-core/molecule/ket_commons.h:43]: Maybe express that this is a constant by naming it ALL_CAPS: `CID` ?

File: core/indigo-core/molecule/ket_commons.h [hunk 10]
```diff
@@ -61,18 +62,18 @@ namespace indigo
         Rect2f _rect;
     };
 
-    class KETTextObject : public GraphMetaObject
+    class KETTextObject : public MetaObject
     {
     public:
         static const std::uint32_t cid = "KET text object"_hash;
```
Comment (SunFellow) [inline core/indigo-core/molecule/ket_commons.h:68]: Maybe express that this is a constant by naming it ALL_CAPS: `CID` ?

File: core/indigo-core/molecule/ket_commons.h [hunk 11]
```diff
@@ -98,5 +99,106 @@ namespace indigo
         std::list<KETTextLine> _block;
         Vec3f _pos;
     };
+
+    class KETReactionArrow : public MetaObject
+    {
+    public:
+        static const std::uint32_t cid = "KET reaction arrow"_hash;
```
Comment (SunFellow) [inline core/indigo-core/molecule/ket_commons.h:106]: Maybe express that this is a constant by naming it ALL_CAPS: `CID` ?

File: core/indigo-core/molecule/ket_commons.h [hunk 12]
```diff
@@ -98,5 +99,106 @@ namespace indigo
...
+
+        int _arrow_type;
+        Vec2f _begin;
+        Vec2f _end;
+    };
+
+    class KETReactionPlus : public MetaObject
+    {
+    public:
+        static const std::uint32_t cid = "KET reaction plus"_hash;
```
Comment (SunFellow) [inline core/indigo-core/molecule/ket_commons.h:138]: Maybe express that this is a constant by naming it ALL_CAPS: `CID` ?

File: core/indigo-core/molecule/molecule_json_saver.h [hunk 13]
```diff
@@ -43,6 +43,7 @@ namespace indigo
         void saveMolecule(BaseMolecule& bmol);
         void saveMolecule(BaseMolecule& bmol, rapidjson::Writer<rapidjson::StringBuffer>& writer);
         bool _add_stereo_desc;
+        static void saveSimpleObjects(const PtrArray<MetaObject>& meta_objects, rapidjson::Writer<rapidjson::StringBuffer>& writer);
```
Comment (SunFellow) [inline core/indigo-core/molecule/molecule_json_saver.h:46]: For the sake of symmetricity, it's better use `MetaObjectsInterface&` here - the same typ[e as in `loadSimpleObjects`.

File: core/indigo-core/reaction/base_reaction.h [hunk 14]
```diff
@@ -64,19 +64,39 @@ namespace indigo
...
         };
 
         BaseReaction();
         virtual ~BaseReaction();
 
+        bool isMultistep()
+        {
+            return _reactionBlocks.size();
+        }
+
```
Comment (SunFellow) [inline core/indigo-core/reaction/base_reaction.h:99]: It looks like never used.

File: core/indigo-core/reaction/src/reaction_json_loader.cpp [hunk 15]
```diff
@@ -32,7 +34,32 @@ using namespace rapidjson;
 
 IMPL_ERROR(ReactionJsonLoader, "reaction KET loader");
 
-ReactionJsonLoader::ReactionJsonLoader(Document& ket) : _rgroups(kArrayType), _molecule(kArrayType), _pluses(kArrayType), _arrows(kArrayType)
+void ReactionJsonLoader::getMoleculeBoundingBox(BaseMolecule& mol, Rect2f& bbox)
```
Comment (SunFellow) [inline core/indigo-core/reaction/src/reaction_json_loader.cpp:37]: It seems can be declared static (unrelated to an instance)

File: core/indigo-core/reaction/src/reaction_json_saver.cpp [hunk 16]
```diff
@@ -204,30 +268,55 @@ void ReactionJsonSaver::saveReaction(BaseReaction& rxn)
...
+        arrow.AddMember("type", "arrow", ket.GetAllocator());
+        Value data(kObjectType);
+        Value pos_array(kArrayType);
+        Value pos1(kObjectType);
+        Value pos2(kObjectType);
+        pos1.AddMember("x", Value().SetDouble(p1.x), ket.GetAllocator());
+        pos1.AddMember("y", Value().SetDouble(-p1.y), ket.GetAllocator());
+        pos1.AddMember("z", Value().SetDouble(0.0), ket.GetAllocator());
+        pos2.AddMember("x", Value().SetDouble(p2.x), ket.GetAllocator());
+        pos2.AddMember("y", Value().SetDouble(-p2.y), ket.GetAllocator());
```
Comment (SunFellow) [inline core/indigo-core/reaction/src/reaction_json_saver.cpp:282]: Why `-p2.y` ?

File: core/render2d/src/render_item_fragment.cpp [hunk 17]
```diff
@@ -58,6 +58,32 @@ void RenderItemFragment::init()
...
+        }
+        break;
+        case KETTextObject::cid:
+            break;
+        case KETReactionArrow::cid:
+            break;
+        case KETReactionPlus::cid:
+            break;
+        default:
+            throw Error("Undefined KET type");
```
[the remark this answers is not in the document]
Comment (even1024) [reply core/render2d/src/render_item_fragment.cpp]: Fixed

File: core/indigo-core/reaction/src/reaction_json_loader.cpp (hunk 15 above)
Comment (even1024) [reply core/indigo-core/reaction/src/reaction_json_loader.cpp]: method moved to BaseMolecule

File: core/indigo-core/reaction/src/reaction_json_saver.cpp [hunk 18]
```diff
@@ -204,30 +268,55 @@ void ReactionJsonSaver::saveReaction(BaseReaction& rxn)
...
         }
+
+        Value arrow(kObjectType);
+        arrow.AddMember("type", "arrow", ket.GetAllocator());
+        Value data(kObjectType);
+        Value pos_array(kArrayType);
+        Value pos1(kObjectType);
+        Value pos2(kObjectType);
+        pos1.AddMember("x", Value().SetDouble(p1.x), ket.GetAllocator());
+        pos1.AddMember("y", Value().SetDouble(-p1.y), ket.GetAllocator());
```
[the remark this answers is not in the document]
Comment (even1024) [reply core/indigo-core/reaction/src/reaction_json_saver.cpp]: fixed

File: core/indigo-core/molecule/ket_commons.h (hunk 11 above)
Comment (even1024) [reply core/indigo-core/molecule/ket_commons.h]: fixed

File: core/indigo-core/molecule/molecule_json_saver.h (hunk 13 above)
Comment (even1024) [reply core/indigo-core/molecule/molecule_json_saver.h]: fixed

File: core/indigo-core/molecule/ket_commons.h (hunk 12 above)
Comment (even1024) [reply core/indigo-core/molecule/ket_commons.h]: fixed

**Review by SunFellow — APPROVED** [review]

**Pull request merged**